File: Java Kompics/src/se/kth/ii2202/experiments/common/PerfectUnicast.java
package se.kth.ii2202.experiments.common;

import java.util.Set;
import se.kth.ii2202.experiments.beb.BEBInit;
import se.sics.kompics.ComponentDefinition;
import se.sics.kompics.Handler;
import se.sics.kompics.Negative;
import se.sics.kompics.Positive;
import se.sics.kompics.address.Address;
import se.sics.kompics.network.Network;
import se.sics.kompics.performancetest.scenario.TopologyView;

/**
 *
 * @author carbone
 */
public class PerfectUnicast extends ComponentDefinition {

    private Negative<PerfectPointToPointLink> pp2p = provides(PerfectPointToPointLink.class);
    private Positive<Network> network = requires(Network.class);
    private Address selfAddr;
    private TopologyView topologyView;
    private Set<Address> neighbors;

    public PerfectUnicast() {
        subscribe(onInit, control);
        subscribe(onSendRequest, pp2p);
        subscribe(onMessageReceived, network);
    }
    Handler<Pp2pInit> onInit = new Handler<Pp2pInit>() {
        @Override
        public void handle(Pp2pInit e) {
            PerfectUnicast.this.topologyView = e.getTopology();
            PerfectUnicast.this.selfAddr = topologyView.getSelfAddress();
        }
    };
    Handler<Pp2pSend> onSendRequest = new Handler<Pp2pSend>() {
        @Override
        public void handle(Pp2pSend e) {
        	//System.out.println("Sending message of type " + e.getMessage().getClass().getCanonicalName());
			trigger(new Pp2pMessage(selfAddr, e.getDestination(), e.getMessage()), network);
        }
    };
    Handler<Pp2pMessage> onMessageReceived = new Handler<Pp2pMessage>() {
        @Override
        public void handle(Pp2pMessage e) {
            trigger(e.getMsg(), pp2p);
        }
    };
}

File: Scala Kompics/src/se/sics/kompics/scala/performancetest/kompics/PerfectUnicast.scala
package se.sics.kompics.scala.performancetest.kompics

import se.sics.kompics.scala.ComponentDefinition
import se.sics.kompics.network.Network
import se.sics.kompics.address.Address
import se.sics.kompics.performancetest.launcher.ScalaTopology
import se.sics.kompics.performancetest.launcher.ScalaTopology._
import se.sics.kompics.performancetest.scenario.TopologyView
import se.sics.kompics.scala.Init
import se.sics.kompics.network.Message
import se.sics.kompics.network.Transport

case class Pp2pMessage(src: Address, dest: Address, msg: Any) extends Message(src, dest, Transport.TCP)

class PerfectUnicast extends ComponentDefinition {
	val pp2p = ++(PerfectPointToPointLink);
	val network = --(classOf[Network]);
	
	private var topology : TopologyView = null;
	
	def self : Address = topology.getSelfAddress();
	def neighbours : Set[Address] = topology.neighbours;

	ctrl uponEvent {case Init(topology: TopologyView) => {() =>
		this.topology = topology;
	}}
	
	pp2p uponEvent {case Pp2pSend(dest, msg) => {() =>
		trigger(Pp2pMessage(self, dest, msg), network);
	}}
	
	network uponEvent {case Pp2pMessage(src, dest, msg) => {() => 
		trigger(Pp2pDeliver(src, msg), pp2p);
	}}
}